internal/dag/node.go

```diff
@@ -32,29 +33,36 @@ func Not(condition NodeCondition) NodeCondition {
 }
 
 // GatherMatchingInputNames scans all the inputs and returns those which match the condition.
-//
-// If direct input doesn't match a condition, search continues up until matching node is found.
 func GatherMatchingInputNames(node Node, condition NodeCondition) []string {
-	result := []string{}
+	return slices.Map(GatherMatchingInputs(node, condition), func(input Node) string {
+		return input.Name()
+	})
+}
+
+// GatherMatchingInputs scans all the inputs and returns those which match the condition.
+func GatherMatchingInputs(node Node, condition NodeCondition) []Node {
+	result := []Node{}
```

Comment:
Something like `return slices.Filter(node.Inputs(), func(input Node) bool { return condition(input) })` should do the trick.